File: Design-Patterns-Using-JavaScript/Basics OOPS/OperatorOverloading.js
class Rectangle {
    constructor(length, breadth) {
        this.length = length;
        this.breadth = breadth;
    }

    area() {
        return this.length * this.breadth;
    }

    perimeter() {
        return 2 * (this.length + this.breadth);
    }

    // Overloading + operator
    add(rectangle) {
        return new Rectangle(this.length + rectangle.length, this.breadth + rectangle.breadth);
    }

    // Overloading - operator
    subtract(rectangle) {
        return new Rectangle(Math.abs(this.length - rectangle.length), Math.abs(this.breadth - rectangle.breadth));
    }

    // Overloading == operator
    equals(rectangle) {
        return this.length === rectangle.length && this.breadth === rectangle.breadth;
    }

    // Overriding toString method
    toString() {
        return `Rectangle length and width: ${this.length} ${this.breadth}`;
    }
}

// Client code
const r1 = new Rectangle(4, 6);
const r2 = new Rectangle(10, 6);

console.log(`Is r1 == r2 ? ${r1.equals(r2)}`);

const r3 = r1.add(r2);
const r4 = r1.subtract(r2);

console.log(r1.toString());
console.log(r2.toString());
console.log(r3.toString());
console.log(r4.toString());

/*
Is r1 == r2 ? false
Rectangle length and width: 4 6
Rectangle length and width: 10 6
Rectangle length and width: 14 12
Rectangle length and width: 6 0
*/

File: Design-Patterns-Using-Kotlin/Basics OOPS/OperatorOverloading.kt
import kotlin.math.abs

open class Rectangle(private val length: Double, private val breadth: Double) {
    fun area(): Double {
        return length * breadth
    }

    fun perimeter(): Double {
        return 2 * (length + breadth)
    }

    // Overloading + operator
    fun add(rec: Rectangle): Rectangle {
        return Rectangle(abs(length + rec.length), this.breadth + rec.breadth)
    }

    // Overloading - operator
    fun subtract(rec: Rectangle): Rectangle {
        return Rectangle(abs(this.length - rec.length), abs(this.breadth - rec.breadth))
    }

    // Overloading == operator
    fun equals(rec: Rectangle): Boolean {
        return this.length == rec.length && this.breadth == rec.breadth
    }

    // Overriding toString method
    override fun toString(): String {
        return "Rectangle length and width: $length $breadth"
    }
}

fun main() {
    val r1 = Rectangle(4.0, 6.0)
    val r2 = Rectangle(10.0, 6.0)

    println("Is r1 == r2 ? " + r1.equals(r2))

    val r3: Rectangle = r1.add(r2)
    val r4: Rectangle = r1.subtract(r2)

    println(r1)
    println(r2)
    println(r3)
    println(r4)
}

/*
Is r1 == r2 ? false
Rectangle length and width: 4.0 6.0
Rectangle length and width: 10.0 6.0
Rectangle length and width: 14.0 12.0
Rectangle length and width: 6.0 0.0
*/